File: generated-code/rust/codecraft/FileReadWrite/src/model/move_action.rs
use super::*;

/// Move action
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct MoveAction {
    /// Target position
    pub target: Vec2I32,
    /// Whether to try find closest position, if path to target is not found
    pub find_closest_position: bool,
    /// Whether to destroy other entities on the way
    pub break_through: bool,
}

impl trans::Trans for MoveAction {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.target.write_to(writer)?;
        self.find_closest_position.write_to(writer)?;
        self.break_through.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let target: Vec2I32 = trans::Trans::read_from(reader)?;
        let find_closest_position: bool = trans::Trans::read_from(reader)?;
        let break_through: bool = trans::Trans::read_from(reader)?;
        Ok(Self {
            target,
            find_closest_position,
            break_through,
        })
    }
}

File: generated-code/pascal/codecraft/FileReadWrite/Model/MoveAction.pas
unit MoveAction;

{$mode delphi}{$H+}

interface

uses
    Stream,
    SysUtils,
    Vec2Int32 in 'Vec2Int32.pas';

type
    // Move action
    TMoveAction = class
        // Target position
        target: TVec2Int32;
        // Whether to try find closest position, if path to target is not found
        findClosestPosition: Boolean;
        // Whether to destroy other entities on the way
        breakThrough: Boolean;
        constructor Create(target: TVec2Int32; findClosestPosition: Boolean; breakThrough: Boolean);
        // Read MoveAction from input stream
        class function ReadFrom(stream: TStream): TMoveAction; static;
        // Write MoveAction to output stream
        procedure WriteTo(stream: TStream);
        function ToString: ansistring; override;
    end;

implementation

constructor TMoveAction.Create(target: TVec2Int32; findClosestPosition: Boolean; breakThrough: Boolean);
begin
    self.target := target;
    self.findClosestPosition := findClosestPosition;
    self.breakThrough := breakThrough;
end;

class function TMoveAction.ReadFrom(stream: TStream): TMoveAction;
var breakThrough: Boolean;
var findClosestPosition: Boolean;
var target: TVec2Int32;
begin
    target := TVec2Int32.ReadFrom(stream);
    findClosestPosition := stream.ReadBoolean;
    breakThrough := stream.ReadBoolean;
    result := TMoveAction.Create(target, findClosestPosition, breakThrough);
end;

procedure TMoveAction.WriteTo(stream: TStream);
begin
    target.WriteTo(stream);
    stream.WriteBoolean(findClosestPosition);
    stream.WriteBoolean(breakThrough);
end;

function TMoveAction.ToString: ansistring;
begin
    result := 'MoveAction {';
    result += 'target=';
    result += target.ToString;
    result += ', ';  
    result += 'findClosestPosition=';
    result += BoolToStr(findClosestPosition);
    result += ', ';  
    result += 'breakThrough=';
    result += BoolToStr(breakThrough);
    result += '}';
end;

end.